pkgs/native_assets_builder/lib/src/dependencies_hash_file/dependencies_hash_file.dart

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:crypto/crypto.dart';
+
+import '../utils/file.dart';
+import '../utils/uri.dart';
+
+class DependenciesHashFile {
+  DependenciesHashFile({
+    required File file,
+  }) : _file = file;
+
+  final File _file;
+  FileSystemHashes _hashes = FileSystemHashes();
+
+  Future<void> _readFile() async {
+    if (!await _file.exists()) {
+      _hashes = FileSystemHashes();
+      return;
+    }
+    final jsonObject =
+        (json.decode(utf8.decode(await _file.readAsBytes())) as Map)
+            .cast<String, dynamic>();
+    _hashes = FileSystemHashes.fromJson(jsonObject);
+  }
+
+  void _reset() => _hashes = FileSystemHashes();
+
+  /// Populate the hashes and persist file with entries from
+  /// [fileSystemEntities].
+  ///
+  /// If [validBeforeLastModified] is provided, any entities that were modified
+  /// after [validBeforeLastModified] will get a dummy hash so that they will
+  /// show up as outdated. If any such entity exists, its uri will be returned.
+  Future<Uri?> hashFiles(
+    List<Uri> fileSystemEntities, {
+    DateTime? validBeforeLastModified,
+  }) async {
+    _reset();
+
+    Uri? modifiedAfterTimeStamp;
+    for (final uri in fileSystemEntities) {
+      int hash;
+      if (validBeforeLastModified != null &&
+          (await uri.fileSystemEntity.lastModified())
```

Comment:
In `flutter_tools` the native assets builder is run in a `Target` concurrently with other `Target`s. Sync I/O prevents flutter_tools making progress on the other targets.

I had to change some sync I/O to async I/O in flutter_tools in other `Target`s a while back to speed up flutter builds because of this.